tarantool_test.go

```diff
@@ -3972,6 +3972,95 @@ func TestConnect_context_cancel(t *testing.T) {
 	}
 }
 
+func TestConnectIsBlocked(t *testing.T) {
+	const server = "127.0.0.1:3015"
+
+	testDialer := dialer
+	testDialer.Address = server
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 10*time.Second)
+	defer cancel()
+	delay := 2 * time.Second
+	errors := make(chan error)
+	go func() {
+		now := time.Now()
+		reconnectOpts := opts
+		reconnectOpts.Reconnect = 100 * time.Millisecond
+		reconnectOpts.MaxReconnects = 100
+		conn, err := Connect(ctx, testDialer, reconnectOpts)
+		if err != nil {
+			errors <- fmt.Errorf("Connection was not established: %v", err)
+			return
+		}
+		defer conn.Close()
+		if time.Now().Before(now.Add(delay)) {
+			errors <- fmt.Errorf("Connection was established too early")
+			return
+		}
+		close(errors)
+	}()
+
+	time.Sleep(delay)
+	inst, err := test_helpers.StartTarantool(test_helpers.StartOpts{
+		Dialer:       testDialer,
+		InitScript:   "config.lua",
+		Listen:       server,
+		WaitStart:    100 * time.Millisecond,
+		ConnectRetry: 10,
+		RetryTimeout: 500 * time.Millisecond,
+	})
+	defer test_helpers.StopTarantoolWithCleanup(inst)
+	if err != nil {
+		t.Fatalf("Unable to start Tarantool: %s", err)
+	}
+	if err := <-errors; err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestConnectIsBlockedUntilContextExpires(t *testing.T) {
+	const server = "127.0.0.1:3015"
+
+	testDialer := dialer
+	testDialer.Address = server
+
+	now := time.Now()
+	delay := 2 * time.Second
+	ctx, cancel := context.WithTimeout(context.TODO(), delay)
+	defer cancel()
+	reconnectOpts := opts
+	reconnectOpts.Reconnect = 100 * time.Millisecond
+	reconnectOpts.MaxReconnects = 100
+	_, err := Connect(ctx, testDialer, reconnectOpts)
+	if err == nil {
+		t.Fatal("Connection was unexpectedly established.")
+	}
+	if time.Now().Before(now.Add(delay)) {
+		t.Fatal("Connect() was unblocked too early")
+	}
+}
+
+func TestConnectIsUnblockedAfterMaxAttempts(t *testing.T) {
+	const server = "127.0.0.1:3015"
+
+	testDialer := dialer
+	testDialer.Address = server
+
+	now := time.Now()
+	ctx, cancel := context.WithTimeout(context.TODO(), 5*time.Second)
```

Comment:
And here.

Reply:
I removed delay and `too early` check. Also I change `MaxReconnects` to 1.